### ISSUE TITLE
Best way to debug OC packages

### ISSUE THREAD

**Issue by alberthajdu** [OrchardCMS/OrchardCore#1521]
What's the best way to debug OC packages when used from the MyGet source (https://www.myget.org/gallery/orchardcore-preview)?
I've found this symbol server doc http://docs.myget.org/docs/reference/symbols but it seems like the orchardcore-preview feed doesn't have a symbol source.

A documentation regarding this topic would be useful too.

**Comment by alberthajdu** [comment]
@sebastienros Any thoughts on this? This could really boost the developer experience and could help identify bugs more quickly.

**Comment by sfmskywalker** [comment]
I'm also very interested in this. Currently I have to basically copy my custom modules to the OrchardCore full-source solution and troubleshoot from there. It works, but increasingly inefficient as our application's complexity increases.

**Comment by jtkech** [comment]
I think we just need to include `pdb` files when packing modules. There is this line in `NuGet.Build.Tasks.Pack.targets`.

    <AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder>.pdb; .mdb; $(AllowedOutputExtensionsInPackageBuildOutputFolder); $(AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder)</AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder>

So maybe we just need to set this property (with a very long name) to true. [See here](https://github.com/NuGet/NuGet.Client/pull/1971) where @jnm2 suggests a shorter property name `IncludeSymbolsInPackage`.

I will look into this tomorrow.

**Comment by jnm2** [comment]
Beyond including the pdbs in the nupkg, you probably also want to GitLink (or SourceLink or embed) your source files if you aren't already. That makes it so much easier for people to step into source- they don't need to have the source on their hard disk and browse to each file. You might be interested in [this comparison](https://github.com/nunit/nunit/issues/2421#issuecomment-357849141) of the available options, especially the final nupkg size comparison lower down.

**Comment by sfmskywalker** [comment]
That would be a dream come true.

**Comment by jtkech** [comment]
I have a solution that works, at least on my machine, i will commit it soon so that you can try it.
@jnm2 thanks for pointing me in the right direction, right now i opted to embed source files.

**Comment by sebastienros** [comment]
MVC uses SourceLink

**Comment by alberthajdu** [comment]
@jtkech you are awesome. One more thing I think it would be useful to mention it in the docs so people would be aware of this thing. And it's basic but it could be helpful if we mention how to enable it (uncheck "Enable Just My Code" in the debugging options).

**Comment by jnm2** [comment]
@alberthajdu Library consumers using the new csproj format will also have to [make sure the PDBs are copied](https://github.com/dotnet/sdk/issues/1458#issuecomment-344422574) to the build output directory for the time being, in addition to unchecking Enable Just My Code. (I highly recommend [Just My Code Toggle](https://marketplace.visualstudio.com/items?itemName=SamHarwell.JustMyCodeToggle) by Sam Harwell if you switch back and forth a lot.)

Basically, everything in https://github.com/nunit/docs/wiki/Debugger-Source-Stepping except step 2 which is specific to GitLink.

**Comment by jnm2** [comment]
Ah, so `DebugType` `embedded` will mean it will work for consumers without them having to do the [PDB copy workaround](https://github.com/dotnet/sdk/issues/1458#issuecomment-344422574). It also means that applications that depend on OC will have no choice but to distribute all that PDB and source inside the OC dll with their app. I think typically people prefer not to deploy debug symbols along with their application.

**Comment by jtkech** [comment]
Good point thanks. I naively thought it could not be an issue. Are you talking about dll sizes and / or performance? I didn't see any too big increase of sizes and no impact on performance.

**Comment by alberthajdu** [comment]
@jtkech the mentioned options are checked for me. I think they are checked by default.

**Comment by jnm2** [comment]
@jtkech It will only affect DLL size. (IP rights aren't an issue since this is open source. 😊) What was the final size? For NUnit, it would have been 3.8 times the original size and we weren't impressed with that.

**Comment by jtkech** [comment]
@alberthajdu okay thanks, so the only difference is that i didn't need to uncheck `Just My Code`, maybe because, for testing, packages were generated on my machine and outputed in a local feed.

@jnm2 okay cool. For most of the libraries it was less than twice, sometimes a little more. After analyzing the sizes of dll, pdb, and nupkg, it seems that a pdb is already compressed when it is embedded in a dll.

**Comment by scil** [comment]
There is a non-best but effecient way https://github.com/OrchardCMS/OrchardCore/discussions/10603
